### ISSUE TITLE
OL 5.0.2 cannot be compiled anymore with angular-cli

### ISSUE THREAD

**Issue by cvenzin** [openlayers/openlayers#8357]
Version 5.0.0 is fine, the new 5.0.2 Version throws the following error during production build for multiple ol modules:

ERROR in ./node_modules/ol/interaction/Pointer.js
Module build failed: Error: Debug Failure. Found typedef tag attached to node which it should not be!
    at nameForNamelessJSDocTypedef ...

**Comment by ahocevar** [comment]
Interesting. Maybe you have changed something in your build confit along the way? Because there were no code changes between 5.0.0 and 5.0.2. See https://github.com/openlayers/openlayers/compare/v5.0.0...v5.0.2. Or maybe it‘s because of the new sideEffects setting in our package.json?

**Comment by ahocevar** [comment]
Can you elaborate a bit on your build stack?

**Comment by cvenzin** [comment]
The build stack is from the newest angular-cli without any modifications. The stack trace is coming out of node_modules\@angular-devkit\build-optimizer when running the production build, the debug build is fine. I don't have the time to dig deeper into the angular build process.
However, I prefer a solution without modifying the angular-cli build process.

**Comment by ahocevar** [comment]
Can you please test if the latest package.json from master (with „sideEffects“: false instead of „sideEffects“: „false“) works?

**Comment by cvenzin** [comment]
yes, i will report on monday. with google i could find similar reprted issues with other packages using this property, probably this bug will be fixed by a future angular-cli version.

**Comment by cvenzin** [comment]
"sideEffects": false also does not compile. However when the build once succeeds with "sideEffects" missing or set to true, the following builds surprisingly also succeed when "sideEffects" is set to false or "false", leading to smaller bundle sizes. Seems to be an issue with the angular-cli after an initial npm install ol ...

**Comment by h3imdall** [comment]
I have the same bug with angular/cli 6.0.8 and openlayers 5.0.3. 
The error only occurs in the production build. 
After setting the property "sideEffects" in the openlayers package.json from false to true I could execute the production build. Then I set the property to false again and no more error messages appear during the production build.

**Comment by DarthPinguin** [comment]
@h3imdall  worked, but it's annoying...

**Comment by martin-reiterer** [comment]
This issue keeps present in openlayers version 5.1.3. Especially in automated build environments, setting the `sideEffects` flag of `package.json` is not always an option, because the lib needs to be forked for adapting the sources.

**Comment by h3imdall** [comment]
@martin-reiterer @DarthPinguin 
I could build openLayers version 5.0.3 with sideeffects: false but i still got bugs when i wanted to open the pages using openLayers, not in dev build but in production build. 
E.g. getLayer is not a function of undefined and so on. 

However I searched the issues and found this. #7879 
It mentions that openlayers is using ES6 in the javascript files. Apparently webpack and or uglifyJs can't handle ES6 in javascript files. 
My workaround is to include the openlayers library in the html header using a cdn and declare the variable ol as any in my angular components. The cdn builds are all built in ES5.
I am not satisfied with it and it is very dirty but it works.

**Comment by ahocevar** [comment]
It looks like this is an angular-cli bug. Has anyone searched or created a ticket there already?

**Comment by martin-reiterer** [comment]
@ahocevar A bug in angular cli was also my first idea, but then I read your comment about the new `sideEffects` flag in `package.json` which addresses tools like webpack. 

The EcmaScript spec specifies that all child modules must be evaluated before they are re-exproted, because they could contain side effects. The `sideEffects` property in `package.json` allows to define your side-effects explicitely. If the package maintainer specifies that the package's modules have no side-effects, then webpack can use this information to optimize re-exports. 

Because of the fact, that setting `sideEffects` to `true` leads to a successful build, I thought that it is more a problem like side-effects within modules of the openlayers lib.

**Comment by ahocevar** [comment]
The error message does not look like it’s actually caused by side effects. It more looks like a typescript error, no? Maybe #8345 will fix it?

**Comment by uweDuesing** [comment]
Hi
I am trying to follow this thread as I am encountering the same problem.  Looking into the package.json of openlayers the parameter   sideEffects looks like this? It is not a boolean, am I looking at the wrong parameter

"sideEffects": [
    "ol.css"
]

**Comment by vmaubert** [comment]
Hi @uweDuesing,

No you're looking to the good parameter: `"sideEffects": true,`

**Comment by hei-pa** [comment]
I tried with `"sideEffects": true` in my project package.json but the error still exists. I am using typescript 2.9.2 with angular 6

**Comment by yangdachu** [comment]
angular-cli : 6.0.8
ol: 5.2.0
ng build is ok, but npm run build error.

![image](https://user-images.githubusercontent.com/23580181/45864540-dd528f00-bdac-11e8-84af-f11bd697ae7e.png)

**Comment by ahocevar** [comment]
This seems to be due to the missing TypeScript declarations. We're working on making OpenLayers TypeScript compatible with JSDoc types. So in future releases, building in angular environments will work again when TypeScript is configured with `allowJs: true`.

**Comment by yangdachu** [comment]
@ahocevar Is there a temporary solution?

**Comment by ahocevar** [comment]
@yangdachu None that I know of.

**Comment by yangdachu** [comment]
@ahocevar  So now I am waiting for your next release, thank you for your help, I wish you a good

**Comment by hei-pa** [comment]
My temporary solution for now is to path the proj.js file before building.

```patch
66,85d65
< 
< /**
<  * A projection as {@link module:ol/proj/Projection}, SRS identifier
<  * string or undefined.
<  * @typedef {module:ol/proj/Projection|string|undefined} ProjectionLike
<  * @api
<  */
< 
< 
< /**
<  * A transform function accepts an array of input coordinate values, an optional
<  * output array, and an optional dimension (default should be 2).  The function
<  * transforms the input coordinate values, populates the output array, and
<  * returns the output array.
<  *
<  * @typedef {function(Array<number>, Array<number>=, number=): Array<number>} TransformFunction
<  * @api
<  */
< 
< 
```

and my package.json 

```json
...
  "scripts": {
    "build": "npm run patchOl && ng build --prod",
    "patchOl": "patch -Ns node_modules/ol/proj.js < patch/ol-typedef.patch > nul || echo Patch for ol already applied"
  }
...
```

**Comment by aalbericio** [comment]
> I have the same bug with angular/cli 6.0.8 and openlayers 5.0.3.
> The error only occurs in the production build.
> After setting the property "sideEffects" in the openlayers package.json from false to true I could execute the production build. Then I set the property to false again and no more error messages appear during the production build.

Worked but it's annoying. Waiting for the new release with this issue fixed...

**Comment by hellfirehd** [comment]
@frankfolsche, GitHub/Markdown has mangled your post.  Would you be willing to provide a gist with your patch file?  I'm at a standstill until this issue is resolved so if your workaround can get me moving again it would be super appreciated.

**Comment by frankfolsche** [comment]
@hellfirehd I've updated the comment to include a gist file, good luck.

**Comment by brandondtran** [comment]
Setting "buildOptimizer" flag to false in Angular.json file will allow for production build. 

This issue is not just specific to OL.  Unfortunately, this problem is rampant in many other libraries including rxjs, angular-tree-component, and tslib just to name few.  Seems like Angular CLI should address this...

**Comment by Chris1234567899** [comment]
I updated all my angular packages (also in dev) to the latest versions and afterwards I could compile it actually. Before that not. Not sure though, which package made the difference in the end.

**Comment by pulsejet** [comment]
Can confirm, everything seems to be working on upgrading to Angular 7

**Comment by vmaubert** [comment]
I confirm, my build is back to normal with the latest angular!

**Comment by arvindshekar07** [comment]
@pulsejet  and @vmaubert  can you please share the setting that you had to make to make angular7 work with ol5.2.0
at present my package.json is like
`json

  "name": “app1”,
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {
    "ng": "ng",
    "start": "ng serve -o",
    "start:proxy": "ng serve -o --proxy-config proxy.conf.json",
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^7.0.0",
    "@angular/common": "^7.0.0",
    "@angular/compiler": "^7.0.0",
    "@angular/core": "^7.0.0",
    "@angular/forms": "^7.0.0",
    "@angular/http": "^7.0.0",
    "@angular/platform-browser": "^7.0.0",
    "@angular/platform-browser-dynamic": "^7.0.0",
    "@angular/router": "^7.0.0",
     "@ngxs/logger-plugin": "^3.2.0",
    "@ngxs/store": "^3.2.0",
    "@turf/turf": "^5.1.6",
    "@types/geojson": "^7946.0.4",
    "@types/long": "^4.0.0",
    "@types/ol": "^4.6.2",
    "@types/turf": "^3.5.32",
    "@webcomponents/custom-elements": "^1.2.1",
    "@webcomponents/shadydom": "^1.2.0",
    "ajv": "^6.5.4",
    "cesium": "^1.50.0",
    "core-js": "^2.5.7",
    "geojson": "^0.5.0",
    "long": "^4.0.0",
    "ol": "^5.2.0",
    "olcs": "^2.4.0",
    "proj4": "^2.5.0",
    "protobufjs": "^6.8.8",
    "rxjs": "^6.3.3",
    "rxjs-compat": "^6.3.3",
    "uuid": "^3.3.2",
    "zone.js": "^0.8.26"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "^0.9.0-rc.3",
    "@angular/cli": "^7.0.0-rc.3",
    "@angular/compiler-cli": "^7.0.0",
    "@angular/language-service": "^7.0.0",
    "@compodoc/compodoc": "^1.1.3",
    "@ngxs/devtools-plugin": "^3.2.0",
    "@types/jasmine": "^2.8.8",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "~10.11.3",
    "canvas-prebuilt": "^1.6.5-prerelease.1",
    "codelyzer": "~4.4.4",
    "jasmine": "^3.2.0”,
    "karma": "^3.0.0",
    "karma-chrome-launcher": "^2.2.0",
    "karma-coverage-istanbul-reporter": "^2.0.4",
    "karma-jasmine": "^1.1.2",
    "karma-jasmine-html-reporter": "^1.3.1",
    "lite-server": "^2.4.0",
    "protractor": "~5.4.1",
    "ts-node": "~7.0.1",
    "tslint": "~5.11.0",
    "typescript": "3.1.1"
  },
  
}
`
Were there any additional changes you had to make to angular.json file and tsconfig.app.json file ?

**Comment by YvainLeyral** [comment]
Hi,
I got the same issue with angular 7.
 @arvindshekar07
 What did you finally changed in your configuration files to solve this issue?

**Comment by yangdachu** [comment]
Hi,

{
  "name": "near",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve --open --port 3000",
    "build": "ng build --prod",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^7.0.3",
    "@angular/cdk": "^7.0.1",
    "@angular/common": "^7.0.3",
    "@angular/compiler": "^7.0.3",
    "@angular/core": "^7.0.3",
    "@angular/forms": "^7.0.3",
    "@angular/http": "^7.0.3",
    "@angular/platform-browser": "^7.0.3",
    "@angular/platform-browser-dynamic": "^7.0.3",
    "@angular/router": "^7.0.3",
    "core-js": "^2.5.4",
    "echarts": "^4.2.0-rc.1",
    "less": "^2.7.3",
    "moment": "^2.22.2",
    "ng-zorro-antd": "^1.8.0",
    "ngx-echarts": "^4.0.1",
    "ol": "^5.3.0",
    "rxjs": "^6.3.3",
    "typescript": "^3.1.6",
    "zone.js": "^0.8.26"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.6.8",
    "@angular/cli": "~7.0.5",
    "@angular/compiler-cli": "^7.0.3",
    "@angular/language-service": "^7.0.3",
    "@types/jasmine": "~2.8.6",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "~8.9.4",
    "codelyzer": "~4.2.1",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~1.7.1",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.0",
    "karma-jasmine": "~1.1.1",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.3.0",
    "ts-node": "~5.0.1",
    "tslint": "~5.9.1"
  }
}


![image](https://user-images.githubusercontent.com/23580181/48905875-ae849200-ee9d-11e8-972c-f2dba74defb9.png)



please help me! think you very much!

**Comment by hajjimurad** [comment]
The issue is still there. My setup:
```
"@angular": "7.0.0",
"@angular/cli": "7.0.5",
"ol": "5.0.2",
```

Only removing of `sideEffects` line in ol/package.json helps

**Comment by pulsejet** [comment]
@hajjimurad you need to update `ol` :)

**Comment by arvindshekar07** [comment]
> Hi,
> I got the same issue with angular 7.
> @arvindshekar07
> What did you finally changed in your configuration files to solve this issue?

Hey sorry for the delayed reply ,  there are some issues wrt to that . On investigating I found that  the when angular builds in --prod mode there is some issue at the time of minimization . Either that is an issue from angular build tools or ol web pack configuration. I think some one has raised issue on in openlayers and angular .

**Comment by mail19** [comment]
"@angular": "7.0.2",
"@angular/cli": "7.0.5",
"ol": "5.3.0",

Removing the property "sideEffects" from /node_modules/ol/package.json was also the fix for me as mentioned by hajjimurad.

**Comment by hajjimurad** [comment]
@mail19 , as I understand the fix will be included in the next version (> 5.3.0)
